examples/tutorials/advanced/Federated Recurrent Neural Network.ipynb

```diff
@@ -0,0 +1,985 @@
+    "    s = now - since\n",
+    "    m = math.floor(s / 60)\n",
+    "    s -= m * 60\n",
+    "    return '%dm %ds' % (m, s)\n",
+    "\n",
+    "def fw_bw_pass_model(model_pointers, line_single, category_single):\n",
+    "    #get the right initialized model\n",
+    "    model_ptr = model_pointers[line_single.location.id]   \n",
+    "    line_reshaped = line_single.reshape(max_line_size, 1, len(all_letters))\n",
+    "    line_reshaped, category_single = line_reshaped.to(device), category_single.to(device)\n",
+    "    #Firstly, initialize hidden layer\n",
+    "    hidden_init = model_ptr.initHidden() \n",
+    "    #And now zero grad the model (not the other way around)\n",
+    "    model_ptr.zero_grad()\n",
+    "    hidden_ptr = hidden_init.send(line_single.location)\n",
+    "    amount_lines_non_zero = len(torch.nonzero(line_reshaped.copy().get()))\n",
+    "    #now need to perform forward passes\n",
+    "    for i in range(amount_lines_non_zero): \n",
+    "        output, hidden_ptr = model_ptr(line_reshaped[i], hidden_ptr) \n",
+    "    criterion = nn.NLLLoss()   \n",
+    "    loss = criterion(output, category_single) \n",
+    "    loss.backward()\n",
+    "    \n",
+    "    model_got = model_ptr.get() \n",
+    "    \n",
+    "    #Perform model weights' updates    \n",
+    "    for param in model_got.parameters():\n",
+    "        param.data.add_(-args.learning_rate, param.grad.data)\n",
+    "        \n",
+    "    model_sent = model_got.send(line_single.location.id)\n",
+    "    model_pointers[line_single.location.id] = model_sent\n",
+    "    \n",
+    "    return(model_pointers, loss, output)\n",
+    "            \n",
+    "  \n",
+    "    \n",
+    "def train_RNN(n_iters, print_every, plot_every, federate_after_n_batches, list_federated_train_loader):\n",
+    "    current_loss = 0\n",
+    "    all_losses = []    \n",
+    "    \n",
+    "    model_pointers = {}\n",
+    "    \n",
+    "    #Send the model to every single worker just before the training procedure starts\n",
+    "    for worker in workers_virtual:\n",
+    "        model_copied = model.copy()\n",
+    "        model_ptr = model_copied.send(worker) \n",
+    "        model_pointers[worker.id] = model_ptr\n",
+    "\n",
+    "    #extract a random element from the list and perform training on it\n",
+    "    for iter in range(1, n_iters + 1):        \n",
+    "        #random_index = random.randint(0, len(list_federated_train_loader) - 1)\n",
+    "        random_index = randomTrainingIndex()\n",
+    "        line_single, category_single = list_federated_train_loader[random_index]\n",
+    "        #print(category_single.copy().get())\n",
+    "        line_name = names_list[random_index]\n",
+    "        model_pointers, loss, output = fw_bw_pass_model(model_pointers, line_single, category_single)\n",
+    "        #model_pointers = fed_avg_every_n_iters(model_pointers, iter, args.federate_after_n_batches)\n",
```

Comment:
So is there already a federated averaging step included? It is commented out here and I didn't see the implementation of the function fed_avg_every_n_iters(). 

For information, there is code to average models already available: see syft.frameworks.torch.federated.utils.federated_avg()

Reply:
Actually, the function fed_avg_every_n_iters performing federated averaging is there in the last commit. It is indeed based on the `utils.federated_avg()` function,  however I commented the federated averaging part because the model loss increases in correspondence with the averaging function call, causing the overall model loss to diverge over time. By discussing this issue with a professor of mine and with Andrew Trask, the reason for that probably the small batch size = 1 being used in the present example.